# SIRFindeR: Intron Retention Estimation and Detection
# Copyright (C) 2019-2020   <lucile.broseus@igh.cnrs.fr>
#########################################################


#_______________________________________________________#
#' Merging SIRratio by reference independent intron.
#'
#' @description This function merges SIRratio values
#' by reference independent intron.
#' 
#' @import magrittr 
#' 
#' @importFrom rlang .data
#' @importFrom data.table fread fwrite
#' @importFrom dplyr select
#'
#' @param saveDir Name for the Sample Directory where files output by computeSIRratio() are saved.
#' @param save Whether the result should be saved in a file (TRUE) or output in R as a \code{data.frame}.
#'
#' @return If save=FALSE, \code{data.frame} with independent intron intervals and their SIRratio.
#' If save=TRUE (default) the \code{data.frame} will be save in \code{saveDir}, in a file called "ResultsByIntron.txt".
#'
#--------------------------------------------------------#

mergebyIntron <- function(saveDir, save = TRUE){
  
  introns <- fread(file = paste(saveDir, "independent_introns.txt", sep = "/"), data.table = F)
  tmp <- fread(file = paste(saveDir, "SIRratio.txt", sep = "/"), data.table = F)
  tmp <- tmp %>% 
    select(intron_group, SpliceLeft, SpliceRight, SpliceMax, intronicCount, SIRratio)
  
  introns <- merge(introns, tmp, by = "intron_group")
  
  if( save ) fwrite(x = introns, file = paste(saveDir, "ResultsByIntron.txt", sep = "/"))
  else return(introns)
  
}

#_______________________________________________________#
#' Intron retention levels estimation using short read RNA-seq data
#'
#' @description This is a wrapper function for computing SIRratio values.
#' It performs all steps required to estimate Intron Retention levels using RNA-seq read alignments and annotation data.
#'
#' @param gtf Reference transcriptome in a gtf/gff file. Used to define introns.
#' @param keepAnnotatedRI Whether intron known as retained in a transcript should be considered (Default: TRUE).
#'
#' @param chainPath (Default: NULL)
#' @param mappabilityFile (eg: wgEncodeCrgMapabilityAlign100mer.bigWig)
#'
#' @param bamFile Path to STAR alignments in a bam file.
#' @param readLength Length of input short reads.
#' @param libraryType Type of the library of RNA-seq data. Either "SE" (single-end) or "PE" (paired-end).
#'
#' @param junctionFile Junction file generated by STAR from the RNA-seq sample.
#'
#' @param min_map_cross_junc Minimal number of supporting reads an intron detected from splice-aware alignments should have (Default: 3).
#' @param min_novel_intron_width Minimal width an intron detected from splice-aware alignments should have (Default: 30).
#'
#'
#' @param saveDir Name for a Project Directory where intermediate and result files will be saved.
#'
#' @param verbose (Default: TRUE)
#'
#' @return A \code{data.frame} with independent intron intervals and their IRratio2.
#'
#' @export
#--------------------------------------------------------#

computeSIRratio <- function(gtf,
                            keepAnnotatedRI = FALSE,

                            chainPath = NULL,
                            mappabilityFile = NULL,

                            bamFile,
                            readLength,
                            libraryType,

                            junctionFile,

                            min_map_cross_junc = 5,
                            min_novel_intron_width = 30,

                            saveDir,
                            verbose = TRUE){

  ## Intron definition:
  defineIntronicIntervals(gtf = gtf,
                          saveDir = saveDir,
                          junctionFile = junctionFile,
                          keepAnnotatedRI = keepAnnotatedRI,
                          min_map_cross_junc = min_map_cross_junc,
                          min_novel_intron_width = min_novel_intron_width,
                          verbose = verbose)

  ## Exclude low mappability intervals:

  minMapScore <- 0.5

  curateIntrons(mappabilityFile = NULL,
                chainPath = NULL,
                readLength = readLength,
                minMapScore = minMapScore,
                saveDir = saveDir)

  ## Extract junction counts per intron_group from a STAR output file

  prepareJunctionSites(gtf,
                       saveDir = saveDir,
                       junctionFile = junctionFile,
                       min_map_cross_junc = min_map_cross_junc)

  ## Get intronic counts from bam file

  appendIntronicCounts(bamFile, saveDir = saveDir, libraryType = libraryType)

  ## Get another estimate of gene expression levels (by gene) by bootstrapping exon to exon counts

  bootstrapJunctionCounts(bamFile, saveDir = saveDir, libraryType = libraryType, nboot = 500)

  ## Merge everything and compute SIRratio

  estimateSIRratio(saveDir, readLength)
  
  ## Merge SIRratio with reference independent introns
  
  mergebyIntron(saveDir, save = T)

}


#_______________________________________________________#
#' Combine all SIRratio values in a same matrix.
#'
#' @description This function creates a matrix of SIRratio
#' with all samples in an experiment.
#' 
#' @import magrittr
#' 
#' @importFrom rlang .data
#' @importFrom data.table fread
#' @importFrom dplyr select distinct
#' 
#' @param saveDirs A vector with all sample directories to be combined.
#'
#' @return A numeric matrix with SIRratio values per samples. 
#' Introns are in rows, samples in columns.
#'
#--------------------------------------------------------#

buildExpSIRmatrix <- function(saveDirs){
  
  files <- list.files(path = saveDirs, pattern = "ResultsByIntron.txt", recursive = T, full.names = T)
  samples <- saveDirs
  
  df <- data.frame()
  
  for(f in 1:length(files)){
    
    tmp <- fread(file = files[f], data.table = F)
    tmp <- tmp %>% dplyr::select(gene_id, seqnames, start, end, strand, SIRratio)
    colnames(tmp)[ncol(tmp)] <- samples[f]
    
    ifelse( f==1, df <- tmp, df <- merge(df, tmp, by = c("gene_id", "seqnames", "start", "end", "strand")))
  }
  df <- df %>% distinct(.keep_all = T)
    
  X <- as.matrix(df[,-c(1:5)])
  rownames(X) <- paste(df$gene_id, "_", df$start, "-", df$end, sep = "")
  
  return( X )
}

#_______________________________________________________#
#' Prepare a matrix of SIRratio values for usual primary analyses
#' (eg: PCA, heatmap) 
#'
#' @description This function processes any SIRratio matrix
#' to make it for for usual primary analyses.
#' Namely, introns with 0 or 1 values will be removed and remaining
#' values will be transformed with the logit function. 
#' 
#' @param X A numeric matrix of SIRratio values per sample (columns)
#'
#' @return A numeric matrix of logit-transformed SIRratio values. 
#'
#--------------------------------------------------------#

prepareSIRmatrix <- function(X){
  
  SIRratioToLogit <- function(x, offset = 0.0001){
    
    if( x<0 ) stop("Input must be positive.")
    
    if( x == 0 ) x <- x + offset
    if( x == 1 ) x <- x - offset
    
    return( log(x/(1-x)) )
    
  }
  
  X <- X[-which(apply(X,1,min)==0),]
  X <- X[-which(apply(X,1,max)==1),]
  
  X <- apply(X, 1:2, SIRratioToLogit)
  
  return( X )
}

