# IRFindeR2: Intron Retention Estimation and Detection
# Copyright (C) 2019-2020   <lucile.broseus@igh.cnrs.fr>
#########################################################


#_______________________________________________________#
#' Intron retention levels estimation using short read RNA-seq data
#'
#' @description This function computes the IRratio2.
#' It performs all steps required to compute
#' the IRratio2 estimator of Intron Retention levels using RNA-seq read alignments and annotation data.
#'
#' @param gtf Reference transcriptome in a gtf/gff file. Used to define introns.
#' @param keepAnnotatedRI Whether intron known as retained in a transcript should be considered (Default: TRUE).
#'
#' @param chainPath (Default: NULL)
#' @param mappabilityFile (eg: wgEncodeCrgMapabilityAlign100mer.bigWig)
#'
#' @param bamFile Path to STAR alignments in a bam file.
#' @param readLength Length of input short reads.
#' @param libraryType Type of the library of RNA-seq data. Either "SE" (single-end) or "PE" (paired-end).
#'
#' @param junctionFile Junction files generated by STAR from the RNA-seq sample.
#' @param sampleJunctionFile Path to the list of exon-exon junctions detected by STAR from read alignments. Used to detect novel introns.
#'
#' @param min_map_cross_junc Minimal number of supporting reads an intron detected from splice-aware alignments should have (Default: 3).
#' @param min_novel_intron_width Minimal width an intron detected from splice-aware alignments should have (Default: 30).
#'
#'
#' @param saveDir Name for a Project Directory where intermediate and result files will be saved.
#'
#' @param verbose (Default: TRUE)
#'
#' @return A \code{data.frame} with independent intron intervals and their IRratio2.
#'
#' @export
#--------------------------------------------------------#

computeIRratio2 <- function(gtf,
                            keepAnnotatedRI = FALSE,

                            chainPath = NULL,
                            mappabilityFile = NULL,

                            bamFile,
                            readLength,
                            libraryType,

                            junctionFile,
                            sampleJunctionFile,

                            min_map_cross_junc = 5,
                            min_novel_intron_width = 30,

                            saveDir,
                            verbose = TRUE){

  ## Intron definition:
  defineIntronicIntervals(gtf = gtf,
                          saveDir = saveDir,
                          junctionFile = junctionFile,
                          keepAnnotatedRI = keepAnnotatedRI,
                          min_map_cross_junc = min_map_cross_junc,
                          min_novel_intron_width = min_novel_intron_width,
                          verbose = verbose)

  ## Exclude low mappability intervals:

  minMapScore <- 0.5

  curateIntrons(mappabilityFile = NULL,
                chainPath = NULL,
                readLength = readLength,
                minMapScore = minMapScore,
                saveDir = saveDir)

  ## Extract junction counts per intron_group from a STAR output file

  prepareJunctionSites(gtf,
                       saveDir = saveDir,
                       junctionFile = sampleJunctionFile,
                       min_map_cross_junc = min_map_cross_junc)

  ## Get intronic counts from bam file

  appendIntronicCounts(bamFile, saveDir = saveDir, libraryType = libraryType)

  ## Get another estimate of gene expression levels (by gene) by bootstrapping exon to exon counts

  bootstrapJunctionCounts(bamFile, saveDir = saveDir, libraryType = libraryType, nboot = 500)

  ## Merge everything and compute IRratio2

  estimateSIRratio(saveDir, readLength)

}
